apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs

```diff
@@ -0,0 +1,161 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Data.Common;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Apis.Auth.OAuth2;
+using Google.Cloud.Spanner.V1;
+
+// ReSharper disable UnusedParameter.Local
+
+namespace Google.Cloud.Spanner
+{
+    /// <summary>
+    /// A connection string builder for Spanner connection strings.
+    /// </summary>
+    public class SpannerConnectionStringBuilder : DbConnectionStringBuilder
+    {
+        private ITokenAccess _credential;
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="connectionString"></param>
+        /// <param name="credential"></param>
+        public SpannerConnectionStringBuilder(string connectionString, ITokenAccess credential) {
+            Credential = credential;
+            base.ConnectionString = connectionString;
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public SpannerConnectionStringBuilder() {
+            
+        }
+
+        internal SpannerConnectionStringBuilder CloneWithNewDataSource(string dataSource) {
+            SpannerConnectionStringBuilder newInstance = new SpannerConnectionStringBuilder();
+            newInstance.Credential = Credential;
+            newInstance.ClientId = ClientId;
+            newInstance.ClientSecret = ClientSecret;
+            newInstance.UserName = UserName;
+            newInstance.DataSource = dataSource;
+            return newInstance;
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public string ClientId
+        {
+            get { return (string)this[nameof(ClientId)]; }
+            private set { this[nameof(ClientId)] = value; }
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public string ClientSecret
```

Comment:
i don't think he fully understands the options I presented.  while clientid and secret are in the connectionstring, there are no refresh tokens or anything else.  It's purely a shortcut for client side auth if you don't use defaultapp credentials.  I do have a proposal for "fixing" auth because right now its either completely obscure magic (default app credentials), or hard to manage (you need to pass around credentials to every api and ensure it has the right scopes).